pkg/schema/loader.go

```diff
@@ -0,0 +1,192 @@
+func ProvideSchemaLoader(goLoader *GoSchemaLoader, themaLoader *ThemaSchemaLoader) *SchemaLoader {
+	return &SchemaLoader{
+		goLoader:    goLoader,
+		themaLoader: themaLoader,
+	}
+}
+
+// LoadSchema loads schema based on schema type and options.
+// It's important that the options for the correct type are provided.
+func (l *SchemaLoader) LoadSchema(
+	ctx context.Context, typ SchemaType, themaOpts ThemaLoaderOpts, goOpts GoLoaderOpts,
+) (ObjectSchema, error) {
+	switch typ {
+	case SchemaTypeThema:
+		return l.themaLoader.LoadSchema(ctx, themaOpts)
+	case SchemaTypeGo:
+		return l.goLoader.LoadSchema(ctx, goOpts), nil
+	default:
+		return nil, ErrUnknownSchemaType
+	}
+}
+
+// GoSchemaLoader loads Go schemas.
+type GoSchemaLoader struct{}
+
+// GoLoaderOpts are the options for loading Go schemas.
+type GoLoaderOpts struct {
+	SchemaName       string
+	GroupName        string
+	GroupVersion     string
+	SchemaOpenapi    apiextensionsv1.JSONSchemaProps
+	SchemaObject     runtime.Object
+	SchemaListObject runtime.Object
+}
+
+// ProvideGoSchemaLoader returns a new GoSchemaLoader.
+func ProvideGoSchemaLoader() *GoSchemaLoader {
+	return &GoSchemaLoader{}
+}
+
+// LoadGoSchema loads a new Go schema with options opts.
+func (l *GoSchemaLoader) LoadSchema(ctx context.Context, opts GoLoaderOpts) GoSchema {
+	return NewGoSchema(
+		opts.SchemaName, opts.GroupName, opts.GroupVersion, opts.SchemaOpenapi, opts.SchemaObject, opts.SchemaListObject,
+	)
+}
+
+// ThemaSchemaLoader loads Thema schemas.
+type ThemaSchemaLoader struct {
+	lib thema.Library
+}
+
+// ThemaLoaderOpts are the options for loading Thema schemas.
+type ThemaLoaderOpts struct {
+	SchemaFS         fs.FS
+	SchemaPath       string
+	SchemaVersion    thema.SyntacticVersion
+	GroupName        string                          // TODO: should be part of the lineage
+	GroupVersion     string                          // TODO: should be part of the lineage
+	SchemaOpenapi    apiextensionsv1.JSONSchemaProps // TODO: should be part of the lineage
+	SchemaType       interface{}
+	SchemaObject     runtime.Object
+	SchemaListObject runtime.Object
+}
+
+// ProvideThemaSchemaLoader returns a new Thema schema loader, which uses Thema library lib for registering schemas.
+func ProvideThemaSchemaLoader(lib thema.Library) *ThemaSchemaLoader {
+	return &ThemaSchemaLoader{
+		lib: lib,
+	}
+}
+
+// LoadThemaSchema loads a new Thema schema with options opts.
+func (l *ThemaSchemaLoader) LoadSchema(ctx context.Context, opts ThemaLoaderOpts) (*ThemaSchema, error) {
+	lin, err := loadThemaLineage(opts.SchemaPath, opts.SchemaFS, l.lib)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calling this ensures our program cannot start,
+	// if the Go DataSource.Model type is not aligned with the canonical schema version in the lineage.
+	if _, err := newJSONKernel(lin, opts.SchemaPath, opts.SchemaType); err != nil {
+		return nil, err
+	}
+
```

Comment:
This can actually be dropped now - all it does internally is call `AssignableTo`, and we're doing that below, so we don't need to rely on this indirect way of doing it anymore.
```suggestion
```

Reply:
Nice, removed!